### DISCUSSION TITLE
Switching to native tools?

### DISCUSSION THREAD

**Discussion by dlants** [olimorris/codecompanion.nvim#494] [Ideas]
I appreciate the efforts to make a provider-agnostic tool system, but I wonder if you've thought about adopting custom protocols for LLMs that start providing native tool support, like [claude](https://www.anthropic.com/news/tool-use-ga).

I played around with it a bit, and it seems fairly robust, so it may improve the ability of the LLM to perform tool use. I imagine they're going to do a fair bit of optimization on their end to make it nicer, compared to a homebrewed solution.

I think you could probably implement a shim to convert from the current xml-based tools to provider-specific tools when it's available... that would probably allow you to keep most of the existing logic intact.

**Comment by olimorris** [comment]
I expect I'll add something like this in early 25 if they're significantly better than the existing tools. As I've said in previous announcements, parity with Aider's features are where I'm trying to get to in the short term.
